core/org.openjdk.jmc.flightrecorder.rules.jdk/src/main/java/org/openjdk/jmc/flightrecorder/rules/jdk/latency/VMOperationRule.java

```diff
@@ -92,36 +97,111 @@ private Result evaluate(IItemCollection items, IQuantity warningLimit) {
+					: Messages.VMOperationRuleFactory_TEXT_WARN;
+			return new Result(this, score, MessageFormat.format(Messages.getString(shortMessage), peakDuration),
+					MessageFormat.format(Messages.getString(longMessage), peakDuration, operation, caller, timeStr),
 					JdkQueries.VM_OPERATIONS_BLOCKING);
 		}
-		return new Result(this, score,
-				MessageFormat.format(Messages.getString(Messages.VMOperationRuleFactory_TEXT_OK), peakDuration), null,
+		String shortMessage = isCombinedDuration ? Messages.VMOperationRuleFactory_TEXT_OK_COMBINED_DURATION
+				: Messages.VMOperationRuleFactory_TEXT_OK;
+		return new Result(this, score, MessageFormat.format(Messages.getString(shortMessage), peakDuration), null,
 				JdkQueries.FILE_READ);
 	}
 
+	private void findLongestEventInfo(IItemCollection items) {
+		IItem startingEvent = null;
+		IQuantity longestDuration = null;
+		IItem curStartingEvent = null;
+		IQuantity prevEndTime = null;
+		IQuantity curCombinedDur = null;
+
+		List<IItem> sortedEvents = sortEventsByStartTime(items);
+		for (IItem event : sortedEvents) {
+			if (curStartingEvent == null) {
+				curStartingEvent = event;
+				curCombinedDur = getDuration(event);
+			} else {
+				IQuantity startTime = getStartTime(event);
+				IQuantity duration = getDuration(event);
+				double timeBetweenEvents = startTime.subtract(prevEndTime).doubleValueIn(UnitLookup.SECOND);
+				if (getOperation(curStartingEvent).equals(getOperation(event))
+						&& getCaller(curStartingEvent).equals(getCaller(event))
+						&& timeBetweenEvents <= MAX_SECONDS_BETWEEN_EVENTS) {
+					curCombinedDur = curCombinedDur.add(duration);
+				} else {
+					curCombinedDur = duration;
+					curStartingEvent = event;
+				}
+			}
+
+			if (longestDuration == null || longestDuration.compareTo(curCombinedDur) < 0) {
+				longestDuration = curCombinedDur;
+				startingEvent = curStartingEvent;
+			}
+			prevEndTime = getEndTime(event);
+		}
+		this.longestDuration = longestDuration;
+		this.startingEvent = startingEvent;
+	}
+
+	private List<IItem> sortEventsByStartTime(IItemCollection items) {
+		List<IItem> sortedEvents = new ArrayList<>();
+		for (IItemIterable iter : items) {
+			for (IItem event : iter) {
+				sortedEvents.add(event);
+			}
+		}
+		Collections.sort(sortedEvents, new Comparator<IItem>() {
+			@Override
+			public int compare(IItem e1, IItem e2) {
+				return getStartTime(e1).compareTo(getStartTime(e2));
+			}
+		});
+		return sortedEvents;
+	}
+
+	private IQuantity getStartTime(IItem event) {
+		return RulesToolkit.getValue(event, JfrAttributes.START_TIME);
+	}
+
+	private IQuantity getEndTime(IItem event) {
+		return RulesToolkit.getValue(event, JfrAttributes.END_TIME);
+	}
+
+	private IQuantity getDuration(IItem event) {
+		return RulesToolkit.getValue(event, JfrAttributes.DURATION);
+	}
+
+	private IMCThread getCaller(IItem event) {
+		return RulesToolkit.getValue(event, JdkAttributes.CALLER);
+	}
```

Comment:
This value can be null.

Reply:
I have made the caller thread comparison account for null values in https://github.com/openjdk/jmc/pull/42/commits/156c9015d78ef12a7b1eaefffbc3aecbf1022967

Reply:
Great!